Create GaussianEnergyFluxBC

lindsayad opened the pull request:
Refs #24462 and #24204

lindsayad committed:
Switch to use FunctionRadiativeBC so we can remove RadiationEnergyFluxBC

lindsayad committed:
Add test for GaussianWeldEnergyFluxBC

lindsayad committed:
Move laser welding heat bcs from malamute into heat conduction

Refs #24462

lindsayad committed:
GaussianWeldEnergyFluxBC doc

lindsayad commented:
This is expected to fail MALAMUTE as I'm migrating this class from there

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1556190/) on 17d3b36 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/24515/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1556231/) on 17d3b36 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>447dca</th>
      <th colspan=4>#24515 17d3b3</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.36%</td>
      <td>85.36%</td>
      <td>-0.00%</td>
      <td>-</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>88359</td>
      <td>88355</td>
      <td>-4</td>
      <td>0</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15150</td>
      <td>15154</td>
      <td>+4</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/24515/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/24515/coverage/framework)


## Modules coverage

### Heat conduction

<table>
  <thead>
    <tr>
      <th></th>
      <th>447dca</th>
      <th colspan=4>#24515 17d3b3</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>88.55%</td>
      <td>88.61%</td>
      <td>+0.06%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>3580</td>
      <td>3603</td>
      <td>+23</td>
      <td>23</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>463</td>
      <td>463</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/24515/coverage_diff/heat_conduction)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/24515/coverage/heat_conduction)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/24515/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/24515/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/24515/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/24515/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/24515/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/24515/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/24515/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/24515/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/24515/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/24515/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/24515/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/24515/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/24515/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/24515/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/24515/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/24515/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/24515/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/24515/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/24515/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/24515/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/24515/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/24515/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/24515/coverage/scalar_transport)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/24515/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/24515/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/24515/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/24515/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/24515/coverage/xfem)

</details>



This comment will be updated on new commits.

GiudGiud commented on modules/heat_conduction/include/bcs/GaussianWeldEnergyFluxBC.h:
```diff
@@ -0,0 +1,30 @@
+
+#include "ADIntegratedBC.h"
+
+class GaussianWeldEnergyFluxBC : public ADIntegratedBC
```
docstring

GiudGiud commented on modules/heat_conduction/src/bcs/GaussianWeldEnergyFluxBC.C:
```diff
@@ -0,0 +1,49 @@
+                              _y_beam_coord.value(_t, _q_point[_qp]),
+                              _z_beam_coord.value(_t, _q_point[_qp])};
+  auto r = (_ad_q_points[_qp] - beam_coords).norm();
+  return -_test[_i][_qp] * 2. * _reff * _F0 * std::exp(-_reff * r * r / (_R * _R));
```
you re going to loose energy from integration error here. Unless you're planning to adapt _F 0?

lindsayad replied:
How would you combat that?

lindsayad replied:
I could just drop this PR and use some parsed function class

lindsayad replied:
> outside of MOOSE, approximating the Gaussian with some polynomial shape that is integrated exactly by the quadrature.

If not done with a lot of thought, this would introduce the exact kind of physics to modeling error

> refining the mesh locally near the laser spot

This is done in the input files using this BC in https://github.com/idaholab/moose/pull/24204 by refining for large gradient jumps (e.g. for large Laplacian) for the temperature field.

lindsayad replied:
There is `FunctionNeumannBC` which I can use in #24204. I don't want to spend a lot of time on this, so if we don't want this BC as written, then I will just move on and close this

GiudGiud replied:
I think the BC is fine. 
I would remove "Weld" from the name though

GiudGiud commented on modules/heat_conduction/test/tests/laser_bc_flux/tests:
```diff
@@ -0,0 +1,10 @@
+    type = Exodiff
+    input = test.i
+    exodiff = test_out.e
+    requirement = 'The system shall be able to model an equilibrium between an incoming heat flux from a laser, which is described by a Guassian shape, and outgoing heat flux due to radiative losses.'
```
```suggestion
    requirement = 'The system shall be able to model an equilibrium between an incoming heat flux from a laser, which is described by a Gaussian shape, and outgoing heat flux due to radiative losses.'
```

GiudGiud replied:
I probably shouldnt be reviewing drafts :S

lindsayad committed:
Remove Weld from class name

lindsayad committed:
Address review

- Class description
- Class doxygen
- Small documentation addition
- Typo fix

GiudGiud commented on modules/heat_conduction/doc/content/source/bcs/GaussianEnergyFluxBC.md:
```diff
@@ -0,0 +1,23 @@
+flux to the centerpoint of the beam. This functional form of the beam flux is
+taken from [!cite](noble2007use). The negative sign on the flux indicates that
+the flux is incoming.
+
```
as a help to the user I would add the integral of the flux assuming all the beam is hitting a flat surface on the domain.

lindsayad replied:
Added a sentence. See if it addresses what you were saying

lindsayad committed:
Use Wikipedia as a reference and update params accordingly

This removes `F_0` and uses `P_0` instead. Moreover, the `r_eff` parameter is
removed.

Also add a surface about beam normal per Guillaume suggestion

lindsayad committed:
Update input parameter values to avoid regolding

lindsayad commented on modules/heat_conduction/doc/content/source/bcs/GaussianEnergyFluxBC.md:
```diff
@@ -0,0 +1,23 @@
+-2r_{eff}F_0\exp{-r_{eff}r_p^2/R^2}
+\end{equation}
+
+where $r_eff$ is an effective radius used to specify the radial distribution of
```
Removed `r_eff`

lindsayad commented on modules/heat_conduction/doc/content/source/bcs/GaussianEnergyFluxBC.md:
```diff
@@ -0,0 +1,23 @@
+\end{equation}
+
+where $r_eff$ is an effective radius used to specify the radial distribution of
+beam energy, $F_0$ is the average heat flux of the beam, $R$ is the beam radius,
```
`F_0` has been changed to be `P_0` for beam power
